Filtering of variant calls produced by metaSNV.py (http://metasnv.embl.de/tutorial.html) in metagenomic samples.  
The script substitutes metaSNV_post.py in the downstream analysis of variant calls.  

Input values should be written to **config.yaml**  

- - - -
## configuration ##
  #### Input files ####
  * SNV_call: Path to called_SNP file generated by metaSNV.  
  * average_coverage: Path to file with average coverage per scaffold generated by metaSNV.  
  * percentage_coverage: Path to file of percentage of coverage by scaffold generated by metaSNV.  
  #### Filtering parameters #####
  * mean_depth_filter: filter on minimal mean depth of coverage. Samples with lower mean of depth in their genome lower than this value are removed.  
  * coverage_filter: filter on minimal coverage of a scaffold. Samples with lower coverage in their genome than this value are removed.  
  #### Others 
  * output_dir: path were output files will be saved  
  * relation:  Relation of the files given. The nucleotide counts observed from samples that share the same relation will be summed up.
    * Example: 1122 
    * Four samples were mapped to the same reference. Two of them belong to one cluster while the other two to another. Order should be the same than the sample's order observed in SNV_call
 - - - -

**RUN**    
python SNV_filter.py  

**Requirements**  
* python 3 or above
* numpy
* yaml
* scipy

**Process**  

SNV_call is opened and readed by row. Each row represents a different variant call in a certain scaffold and position.  
Counts of alternative alleles in each of the avialable samples are used to calculate the frequency of alternative allele per position and per sample. Counts from samples specified to be in the same cluster are summed up (config file, realtion).    
Positions with more than one alternative allele are considered independently (as a totally different position in each of the alleles).  
A matrix of alternative allele frequency per position is generated, scaffolds are assumed to be from the same reference.  
Pairwise Manhattan distances are generated between each of the samples.  
Fst values are calculated between each pair of samples as described in Schloissnig et al., but assuming that multiple alternative alleles are independent and therefore, using the allele frequence of the alternative and 1-allele frequence as the refrence frequency.

1. Costea, Paul Igor, et al. "metaSNV: a tool for metagenomic strain level analysis." PLoS One 12.7 (2017): e0182392.
2. Schloissnig S, Arumugam M, Sunagawa S, Mitreva M, Tap J, Zhu A, et al. Genomic variation landscape of the human gut microbiome. Nature. Nature Publishing Group; 2013;493: 45â€“50. pmid:23222524

