---
title: "Oral_microbiome"
author: "Sergio Andreu-S'{a}nchez"
date: "28 de enero de 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Analysis of Oral Microbiome

In this Rmarkdown document the different steps on the microbiome analysis on Nijmegen oral microbiome project will be presented. The input samples consists on a metadata matrix, a matrix of microbiome relative abundance produced by Metaphlan, a matrix of pathway abundances produced by humann2 and a file with statistics on human contamination and number of reads.

Let us begin by loading the packages needed and setting up the work location where the files are located.
I use tidyverse mainly to use function from dplyr, ggplot and readr, for easy data handling. Vegan and ape are phylogenetic and ecological packages that are widely used in microbiome analysis. Ggforce extends some functions in ggplot. Ggtree is a ggplot-based tree visualizing tool. Cluster allows to cluster the data.

```{r Prepare libraries}
setwd("~/PhD/WORK/Nijmegen/")

suppressWarnings(suppressMessages(library(tidyverse)))
suppressWarnings(suppressMessages(library(vegan)))
suppressWarnings(suppressMessages(library(ape)))
suppressWarnings(suppressMessages(library(ggforce)))
suppressWarnings(suppressMessages(library(ggtree)))
suppressWarnings(suppressMessages(library(cluster)))
```

## Loading data

Next I will load the data using readr functions. The taxonomy matrix contains information on different taxonomic levle encoded as a letter followed by two underscroes. It also contains a second header that I will remove.
The metadata is per patient. As a single patient may have both saliva and pocket tissue, I duplicate the metadata information for both pocket samples, with a P ID and saliva samples, with a S ID.
Finally, pathways file contain both information on individual bacteria and total pathway information, i.e there is the amount of pathway X in bacteria y,z and w and the overall abundance of pathway X. We are just interested in the later, thus we will remove all information on bacteria (after a | character)

```{r Data load, message=F}
matrix_taxonomy = read_tsv("Merged_batch.tsv")
matrix_taxonomy[-1,] -> matrix_taxonomy
matrix_taxonomy %>% filter(grepl("s__",ID)) %>% filter(!grepl("t__",ID))  -> species_matrix
matrix_taxonomy %>% filter(grepl("g__",ID) & !grepl("s__",ID)) -> genus_matrix
matrix_taxonomy %>% filter(grepl("f__",ID) & !grepl("g__",ID)) -> family_matrix


sample_metadata = read_delim("Metadata.csv", delim =";")
sample_metadata %>% mutate(ID_patient = StudieID ,StudieID = paste0(sample_metadata$StudieID,"P"), Source="pocket")-> metadata_P  
sample_metadata %>% mutate(ID_patient = StudieID, StudieID = paste0(sample_metadata$StudieID,"S"), Source="saliva")-> metadata_S  
sample_metadata = rbind(metadata_P,metadata_S)


pathway_abundance = read_tsv("Merged_batch_pathabundance.tsv")
pathway_abundance %>% filter(! grepl("\\|",`# Pathway`) )-> pathway_abundance

```

## Preparing functions

I will first write down the functions I will be calling on the data later.

### Descriptive functions
First, I have two functions that are descriptive on the data. 

"do_Metrics" generates a summary on Path N of a given matrix. It counts for each feature (taxa) the number of 0 we observe and some statistics about them. 
```{r Create metrics on 0s in each taxa, message=F}
do_Metrics = function(MData,N){
  ID = MData[dim(MData)[2]] 
  MData %>% dplyr::select(-one_of(c("ID"))) -> MData
  number_0 = function(V){
    N_0 = 0
    for (i in V){
      if(i == 0){ N_0 = N_0 + 1}
    }
    return(N_0)
  }
  
  Total_summary = tibble()
  for (C in seq(1:dim(MData)[2])){
  MData[,C] %>% mutate_if(is.character, as.numeric) %>% summarise_all(list(min, max,median,mean,sd,number_0)) %>% mutate(Name = colnames(MData[,C])) -> SUMMARY
  colnames(SUMMARY) = c("Minimum_abundance", "Maximum_abundance", "Median_abundance", "Mean_abundance","Standard_deviation","Number_of_0_counts","Organism")
  rbind(Total_summary, SUMMARY) -> Total_summary
  }
  write_csv(x = Total_summary, path = N)
}

```
"contamination_stats" creates some plots regarding contamination. It takes the metadata information and the contamination statistics. It checks on: Contamination vs nReads, Tissue vs Contamination, and the distribution of contamination.


```{r Check contamination statistics, message=F}
contamination_stats = function(File = "Reads_contaminated.txt.txt", metadata=sample_metadata){
  Contaminated_metrics = read_delim(File, delim=" ")
  Contaminated_metrics %>% mutate(Total= Human_reads+Paired_Reads, Cont_percentage=Human_reads/Total) -> Contaminated_metrics
  Contaminated_metrics %>% ggplot() + geom_density(aes(x=Cont_percentage)) + theme_bw() -> contamination_distribution
  ggplot(Contaminated_metrics,aes(x=Total, y=Cont_percentage)) + geom_point() + theme_bw() + geom_smooth(method='lm', formula= y~x) ->contamination_vs_reads

  metadata %>% filter(StudieID %in% Contaminated_metrics$Sample) %>% arrange(StudieID) -> metadata
  Contaminated_metrics %>% filter(Sample %in% metadata$StudieID) %>% arrange(Sample) -> Contaminated_metrics
  Contaminated_metrics %>% mutate(Source = metadata$Source) -> Contaminated_metrics
  Contaminated_metrics %>%
    ggplot(aes(x=Source, y=Cont_percentage)) + geom_boxplot()+ geom_sina()  + theme_bw()
  
  wilcox.test(filter(Contaminated_metrics, Source=="pocket")$Cont_percentage, filter(Contaminated_metrics, Source=="saliva")$Cont_percentage)
  print(contamination_distribution)
  print(contamination_vs_reads)
  
}

```

"distribution_taxonomy", it takes the overall taxonomy matrix and counts the different taxonomic classifications identified, overall comparing the two sources (saliva and pocket), by individual, and whether the number of reads plays a role in the taxons identified.

```{r Check the taxonomy metrics, message=F}
distribution_taxonomy = function(taxa_matrix = matrix_taxonomy){
  t_class = taxa_matrix %>% filter(grepl("t__",ID)) %>% mutate(Level="strain")
  s_class = taxa_matrix %>% filter(grepl("s__",ID)) %>% filter(!grepl("t__",ID)) %>% mutate(Level="species")
  g_class = taxa_matrix %>% filter(grepl("g__",ID) & !grepl("s__",ID)) %>% mutate(Level="genus")
  f_class = taxa_matrix %>% filter(grepl("f__",ID) & !grepl("g__",ID)) %>% mutate(Level="family")
  o_class = taxa_matrix %>% filter(grepl("o__",ID) & !grepl("f__",ID)) %>% mutate(Level="order")
  c_class = taxa_matrix %>% filter(grepl("c__",ID) & !grepl("o__",ID)) %>% mutate(Level="clade")
  p_class = taxa_matrix %>% filter(grepl("p__",ID) & !grepl("c__",ID)) %>% mutate(Level="phylum")
  k_class = taxa_matrix %>% filter(grepl("k__",ID) & !grepl("p__",ID)) %>% mutate(Level="domain")
  
  print("Counting number the species identified per taxonomic level")
  
  raw_numbers = tibble(strain = dim(t_class)[1], species = dim(s_class)[1], 
              genus = dim(g_class)[1], family = dim(g_class)[1], order = dim(o_class)[1],
              clade = dim(c_class)[1], phylum = dim(p_class)[1], domain = dim(k_class)[1])
  raw_numbers %>% gather(Taxonomic_level, Number, strain:domain, factor_key=TRUE) -> tax_numbers
  ggplot(tax_numbers) + geom_bar(aes(x=Taxonomic_level,y=Number), stat="identity") + theme_bw() -> taxonomy_numbers
  print(taxonomy_numbers)
  
  
  print("Counting how many items are there per taxonomic level in each individual")
  classified = tibble()
  for (t_s in list(t_class,s_class,g_class,f_class,o_class,c_class,p_class,k_class)){
    taxa = t_s$ID
    transposed = as_tibble(t(select(t_s,-c(ID,Level))))
    colnames(transposed) = taxa
    transposed  %>% mutate_if(is.character, as.numeric) -> transposed
    (! as.matrix(transposed) == 0)*1 -> transposed
    
    classified_i = tibble()
    for (Indv in seq(1:dim(transposed)[1])){
      presences = sum(as_vector(transposed[Indv,]))
      presences = tibble(N=presences, Level=t_s$Level[1], ID =colnames(t_s)[Indv+1])
      classified_i = rbind(classified_i,presences)
    }
    
    classified = rbind(classified,classified_i)
  }
  
  
  classified %>% mutate(ID= str_replace(ID,"_feature_counts","")) %>% filter(!ID=="Pipelines") ->classified
  
  classified %>% arrange(desc(N)) -> ordered
  classified %>% mutate(Level = factor(Level,levels= c("strain","species","genus","family","order","clade","phylum","domain"))) -> classified
  classified %>% mutate(ID = factor(ID,levels= unique(ordered$ID))) -> classified
  
  classified %>% ggplot(aes(x=ID, y=N)) + geom_bar(stat="identity") + facet_wrap(~Level) + theme_bw() + theme(axis.text.x = element_text(angle = 90)) -> number_classified_by_ID
  print(number_classified_by_ID)
  
  
  print("Loading reads per sample, assesing whether the number of reads affect the taxons identified in each sample")
  Reads = read_delim("Reads_contaminated.txt.txt"," ")
  Reads %>% filter(!Sample== "Pipelines") -> Reads
  Read_number = vector()
  for (entry in arrange(Reads,Sample)$Paired_Reads){
    for (i in c("strain","species","genus","family","order","clade","phylum","domain")){ 
      Read_number = c(Read_number,entry)
    }
  }
  arrange(classified,ID) %>% mutate(Reads= Read_number) -> Classified_reads
  ggplot(Classified_reads, aes(x=N,y=Reads,col=ID)) + geom_point() + facet_wrap(~Level, scales="free") + theme_bw()  + theme(legend.position = "none") -> reads_vs_class
  print(reads_vs_class)
}
```

"do_bars", generates a bar plot comparing the taxonomic relative abundance betweeen the two tissues, and between patients and controls. It also prints the top 5 taxa in each case.

```{r Create bar plots comparing Source and Group composition, message=F}
do_bars = function(taxa_matrix, metadata){
  metadata %>% filter(StudieID %in% taxa_matrix$ID) %>% arrange(StudieID) -> metadata
  taxa_matrix %>% filter(ID %in% metadata$StudieID) %>% arrange(ID) -> taxa_matrix
  taxa_matrix = select(taxa_matrix, -ID)
  
  metadata %>% mutate(N_ID=as.numeric(gsub("[^0-9.-]", "", StudieID))) %>% arrange(N_ID) -> test
  metadata$StudieID = factor(metadata$StudieID,levels(factor(metadata$StudieID))[match(test$StudieID,metadata$StudieID)])
  
  cbind(metadata,taxa_matrix) -> combined_matrix
  
  #tissue
  print("Study taxa composition by tissue")
  combined_matrix %>% gather(Taxa, Abundance, (dim(metadata)[2]+1):dim(combined_matrix)[2], factor_key=TRUE) %>%
  ggplot() + geom_bar(aes(x=StudieID,y=Abundance, fill=Taxa), stat = "identity",position="stack") + facet_wrap(~Source, dir="v")+
   theme_bw() + theme(axis.text.x = element_text(angle = 90), legend.position = "bottom",legend.title = element_text(size = 3), legend.text = element_text(size = 6))+
    theme(legend.key.size = unit(0.1, "cm")) -> BAR1
  print(BAR1)
 
  print("Study taxa composition in Patients vs Controls")
  combined_matrix %>% gather(Taxa, Abundance, (dim(metadata)[2]+1):dim(combined_matrix)[2], factor_key=TRUE) %>%
    ggplot() + geom_bar(aes(x=StudieID,y=Abundance, fill=Taxa), stat = "identity",position="stack") + facet_wrap(~factor(Group), dir="v")+
    theme_bw() + theme(axis.text.x = element_text(angle = 90), legend.position = "bottom",legend.title = element_text(size = 3), legend.text = element_text(size = 6))+
    theme(legend.key.size = unit(0.1, "cm")) -> BAR2
  print(BAR2)
  
  
  print("Top 5 taxons by tissue")
  combined_matrix %>% filter(Source=="saliva") %>% select(colnames(taxa_matrix)) %>% summarise_all(median) %>%
    gather(Taxa, Median_relative_abundance, 1:length(colnames(taxa_matrix)), factor_key=TRUE) %>%
  arrange(desc(Median_relative_abundance)) %>% head(n=5) -> top
   
  combined_matrix %>% filter(Source=="pocket") %>% select(colnames(taxa_matrix)) %>% summarise_all(median) %>%
    gather(Taxa, Median_relative_abundance, 1:length(colnames(taxa_matrix)), factor_key=TRUE) %>%
    arrange(desc(Median_relative_abundance)) %>% head(n=5) -> top2
  
  print("Top 5 taxons by condition")
  
  combined_matrix %>% filter(as.character(Group)=="0") %>% select(colnames(taxa_matrix)) %>% summarise_all(median) %>%
    gather(Taxa, Median_relative_abundance, 1:length(colnames(taxa_matrix)), factor_key=TRUE) %>%
    arrange(desc(Median_relative_abundance)) %>% head(n=5) -> top3
  
  combined_matrix %>% filter(as.character(Group)=="1") %>% select(colnames(taxa_matrix)) %>% summarise_all(median) %>%
    gather(Taxa, Median_relative_abundance, 1:length(colnames(taxa_matrix)), factor_key=TRUE) %>%
    arrange(desc(Median_relative_abundance)) %>% head(n=5) -> top4
  
  top %>% mutate(Top="Saliva") -> top
  top2 %>% mutate(Top="Pocket") -> top2
  top3 %>% mutate(Top="Healthy") -> top3
  top4 %>% mutate(Top="Patient") -> top4
  TOP = rbind(top,top2,top3,top4)
  return(TOP)
  
}
```


### Processing functions

First, we need to change a bit the taxa matrix. It has each bacteria as a row and each sample as column. We want it the other way around. We also call do_metrics on each of the samples that are preprocessed

```{r Preprocess taxa, message=F}
Preprocess_f = function(Mtx,Summary_name){
#Transpose matrix
Sample = gsub("_feature_counts","",colnames(Mtx))
Sample = Sample[2:length(Sample)]

Taxon = vector()
list_split = strsplit(Mtx$ID, "\\|")
for (i in seq(1:length(list_split))){
  element =  list_split[[i]]
  Taxon = c(Taxon, element[length(element)])
}

Mtx[-1,] -> Mtx2
Mtx2 = as_tibble(t(Mtx)[-1,])
Mtx2%>% mutate_if(is.character, as.numeric) -> Mtx2

colnames(Mtx2) = Taxon

do_Metrics(mutate(Mtx2, ID = Sample), N= Summary_name)

taxa2 = Mtx2[2:length(Mtx2)]
taxa_filt= taxa2 

mutate(taxa_filt, ID = Sample) -> taxa_filt
taxa_filt[c(dim(taxa_filt)[2], seq(1:(dim(taxa_filt)[2]-1)))] -> taxa_filt

return(taxa_filt)
}

```

Next, we want to check Alpha-diversity (within samples microbial composition) differences and Beta-diversity (differences between samples) differences.

```{r Diversity, message=F}
Diversity_analysis = function(taxa_matrix,metadata){
  print("Matching data and metadata IDs")
  metadata %>% filter(StudieID %in% taxa_matrix$ID ) %>% arrange(StudieID) -> metadata  
  taxa_matrix %>% filter(ID %in% metadata$StudieID ) %>% arrange(ID) -> taxa_matrix 

  ###Clustering of the data
    ## Hierarchical clustering
  print("Clustering based on disimilarity matrix")
  dist <- vegdist(taxa_matrix[,2:dim(taxa_matrix)[2]],  method = "bray")
  fit = hclust(dist)
  tree = as.phylo(fit)
  tree$tip.label = metadata$StudieID
  ggtree(tree)  %<+%  metadata + geom_tiplab(aes(col=Source),hjust=-0.3) + geom_tippoint(aes(shape=factor(Group)),size=3) + #aes(col=factor(Group)) + 
    xlim(NA,0.7) +  theme(legend.position=c(.9, .6)) -> Cluster_plot
  print(Cluster_plot)
  
  ### Alpha diversity
  print("Computing in sample alpha diversity and plotting")
  shannon_diversity = diversity(taxa_matrix[,2:dim(taxa_matrix)[2]], index = "shannon")

  alpha_diversities =  mutate(metadata, Shannon=shannon_diversity)
  gather(alpha_diversities, Diversity_index, value, c(Shannon), factor_key=TRUE) -> alpha_diversities
  
  alpha_diversities %>% ggplot(aes(x=factor(Group), y=value,col=Source)) + geom_boxplot() +geom_sina() +
  facet_wrap(~Diversity_index, scales= "free") + theme_bw() -> Fig_alpha
  print(Fig_alpha)
  
  ### Beta Diversity
  print("Calculating Beta diversity and computing PCoA")
  dist <- vegdist(taxa_matrix[,2:dim(taxa_matrix)[2]],  method = "bray")
  PCoA = pcoa(dist)

  ggplot(data=as_tibble(PCoA$values)) + geom_bar(aes(x=seq(1:dim(PCoA$values)[1]), y=Relative_eig), col="black", stat = "identity") + 
    theme_bw() -> Fig_Scree_plot
  print(Fig_Scree_plot)
  Eig_vector = PCoA$vectors[,1:3]
  metadata %>% mutate(PCoA1 = Eig_vector[,1], PCoA2 = Eig_vector[,2], PCoA3=Eig_vector[,3]) -> ordination_data
  ggplot(data= ordination_data) + geom_point(aes(x = PCoA1, y=PCoA2, col=Source, shape=factor(Group))) +
    theme_bw() + labs(x=paste(c("PCoA1(",as.character(round(PCoA$values[1,2],3)*100),"%)"), collapse=""), y=paste(c("PCoA2(",as.character(round(PCoA$values[2,2],3)*100),"%)"), collapse="")) -> Fig_PCoA1
  ggplot(data= ordination_data) + geom_point(aes(x = PCoA2, y=PCoA3, col=Source, shape=factor(Group))) +
    theme_bw() + labs(x=paste(c("PCoA2(",as.character(round(PCoA$values[2,2],3)*100),"%)"), collapse=""), y=paste(c("PCoA3(",as.character(round(PCoA$values[3,2],3)*100),"%)"), collapse="")) -> Fig_PCoA2
  print(Fig_PCoA1)
  print(Fig_PCoA2)
  
 ##Getting contamination stats 
  Reads = read_delim("Reads_contaminated.txt.txt",delim=" ")
  metadata %>% mutate(n_reads = arrange(Reads,Sample)$Paired_Reads) -> metadata
  
###Alpha diversity statistical testing
  print("Using wilcoxon rank test on saliva difference on alpha diversity between patients and controls")
  alpha_diversities %>% filter(Source=="saliva") -> Saliva_diversity
  ggplot(data=Saliva_diversity) + geom_density(aes(x=value, col=factor(Group))) + facet_wrap(~Diversity_index, scales="free") + theme_bw()
  for (Diversity in unique(Saliva_diversity$Diversity_index)){
    Saliva_diversity %>% filter(Diversity_index == Diversity) -> MS
    Test = wilcox.test(filter(MS, Group=="0")$value, filter(MS, Group=="1")$value)
    print(c(Diversity,Test$p.value))
  }
  print("Using wilcoxon rank test on pocket difference on alpha diversity between patients and controls")
  alpha_diversities %>% filter(Source=="pocket") -> Pocket_diversity
  ggplot(data=Pocket_diversity) + geom_density(aes(x=value, col=factor(Group))) + facet_wrap(~Diversity_index, scales="free") + theme_bw()
  for (Diversity in unique(Saliva_diversity$Diversity_index)){
    Pocket_diversity %>% filter(Diversity_index == Diversity) -> MS
    Test = wilcox.test(filter(MS, Group=="0")$value, filter(MS, Group=="1")$value)
    print(c(Diversity,Test$p.value))
  }
  #######
  
  ###Beta diversity Statistical testing
  print("Doing Permanova test on Beta diversity")
  permanova_1 = adonis2(dist~ Source + factor(Group) +n_reads,data=metadata, permutations = 999,strata=Source, method="bray")
  
  #Divide by tissue
  metadata %>% filter(Source=="pocket") -> meta_pocket
  metadata %>% filter(Source=="saliva") -> meta_saliva
  taxa_matrix %>% filter(ID %in% meta_saliva$StudieID) -> saliva_matrix
  taxa_matrix %>% filter(ID %in% meta_pocket$StudieID) -> pocket_matrix
  saliva_dist <- vegdist(saliva_matrix[,2:dim(saliva_matrix)[2]],  method = "bray")
  pocket_dist <- vegdist(pocket_matrix[,2:dim(pocket_matrix)[2]],  method = "bray")
  
  permanova_2_saliva = adonis2(saliva_dist~ factor(Group)+n_reads ,data=meta_saliva, permutations = 999,strata=Source, method="bray")
  permanova_2_pocket = adonis2(pocket_dist~ factor(Group)+n_reads ,data=meta_pocket, permutations = 999,strata=Source, method="bray")
  print("Total permanova")
  print(permanova_1)
  print("In saliva permanova")
  print(permanova_2_saliva)
  print("In pocket permanova")
  print(permanova_2_pocket)
  ###########
}
```


Finally, we will look at individual taxa differences

```{r Differential abundance}
Differential_abundance = function(taxa_matrix,metadata, transformation = "argsinsquare"){
  
  metadata %>% filter(StudieID %in% taxa_matrix$ID) %>% arrange(StudieID) -> metadata
  Reads = read_delim("Reads_contaminated.txt.txt",delim=" ")
  metadata %>% mutate(n_reads = arrange(Reads,Sample)$Paired_Reads) -> metadata
  taxa_matrix %>% filter(ID %in% metadata$StudieID) %>% arrange(ID) -> taxa_matrix
  abundance_matrix = taxa_matrix[,2:dim(taxa_matrix)[2]]
  
  print("Filtering Columns absent in over 80% of samples")
  abundance_matrix = abundance_matrix[,((colSums(abundance_matrix !=0) / nrow(abundance_matrix)) *100 )>20]
  
  print("Transforming data")
  if(transformation == "argsinsquare"){ abundance_matrix = asin(sqrt(abundance_matrix/100)) } 
  
  print("Iterating by taxa and fitting a linear model")
  Model_1 = fit_linear_model(abundance_matrix, metadata, "source")

  
  as_tibble(abundance_matrix) %>% mutate(ID=metadata$StudieID) -> temporal_matrix
  metadata %>% filter(Source=="pocket") -> meta_pocket
  metadata %>% filter(Source=="saliva") -> meta_saliva
  temporal_matrix %>% filter(ID %in% meta_pocket$StudieID) %>% select(-ID) -> pocket_matrix
  temporal_matrix %>% filter(ID %in% meta_saliva$StudieID) %>% select(-ID) -> saliva_matrix
  
  print("Iterating by taxa and fitting a linear model on pocket samples")
  Model_2 = fit_linear_model(pocket_matrix, meta_pocket, vector())
  print("Iterating by taxa and fitting a linear model on saliva samples")
  Model_3 = fit_linear_model(saliva_matrix, meta_saliva, vector())

  print(c(dim(filter(Model_1,FDR < 0.1)), dim(filter(Model_2,FDR < 0.1)), dim(filter(Model_3,FDR<0.1))))
  
}

```
And to individual pathway differences
```{r Differential pathway abundance}
differential_pathway_analysis = function(Pathway_abundance, metadata){
  metadata %>% filter(StudieID %in% Pathway_abundance$ID) %>% arrange(StudieID) -> metadata
  Pathway_abundance %>% filter(ID %in% metadata$StudieID) %>% arrange(ID) -> Pathway_abundance
  
  Reads = read_delim("Reads_contaminated.txt.txt",delim=" ")
  metadata %>% mutate(n_reads = arrange(Reads,Sample)$Paired_Reads) -> metadata
  
  
  Abundance_matrix = select(Pathway_abundance, -c(ID))
  #Transformation
  Abundance_matrix = asin(sqrt(Abundance_matrix/100))
  
  
  Model_1 = fit_linear_model(abundance_matrix = Abundance_matrix, metadata = metadata, covariates = "source")
  
  ##Tissue-specific models
  as_tibble(Abundance_matrix) %>% mutate(ID=metadata$StudieID) -> temporal_matrix
  metadata %>% filter(Source=="pocket") -> meta_pocket
  metadata %>% filter(Source=="saliva") -> meta_saliva
  temporal_matrix %>% filter(ID %in% meta_pocket$StudieID) %>% select(-ID) -> pocket_matrix
  temporal_matrix %>% filter(ID %in% meta_saliva$StudieID) %>% select(-ID) -> saliva_matrix
  
  print("Iterating by taxa and fitting a linear model on pocket samples")
  Model_2 = fit_linear_model(pocket_matrix, meta_pocket, vector())
  print("Iterating by taxa and fitting a linear model on saliva samples")
  Model_3 = fit_linear_model(saliva_matrix, meta_saliva, vector())
  as_tibble(Model_2) %>% group_by(FDR<0.01, FDR_normality>0.01) %>% summarise(n()) -> C
  as_tibble(Model_3)%>% group_by(FDR<0.01, FDR_normality>0.01) %>% summarise(n()) -> C2
  as_tibble(Model_1)%>% group_by(FDR<0.01, FDR_normality>0.01) %>% summarise(n()) -> C0
  
  
  return(rbind(mutate(C0, TEST="all"), mutate(C, TEST="pocket"), mutate(C2, TEST="saliva")))
  
}


```



### Helper functions
Iterate and fit linear model
```{r Fit linear model to Taxa/Pathway}
fit_linear_model = function(abundance_matrix,metadata, covariates){
  result_bugg = tibble()
  for (N in seq(1:dim(abundance_matrix)[2])){
    dependent = as_vector(abundance_matrix[,N])
    if (sum(dependent)==0){next
    }else if (sum((!dependent==0)*1)/sum(dependent) < 0.2){ next }
    
    metadata %>% mutate(Dependent = dependent) -> regression_input
    distribution_bug = ggplot(regression_input) + geom_density(aes(Dependent)) + theme_bw() + facet_wrap(~Source)
    
    if ("source" %in% covariates & length(covariates) == 1){ 
      lm(Dependent ~ Source + factor(Group) + n_reads, data= regression_input ) -> regression_output
      S = summary(regression_output)
      fold_change = mean(filter(regression_input, Group == 1)$Dependent) /  mean(filter(regression_input, Group == 0)$Dependent)
      pvalue = as_tibble(S$coefficients)$`Pr(>|t|)`[3]
   }else if(length(covariates)==0){
      lm(Dependent ~ factor(Group) + n_reads, data= regression_input ) -> regression_output
      S = summary(regression_output)
      fold_change = mean(filter(regression_input, Group == 1)$Dependent) /  mean(filter(regression_input, Group == 0)$Dependent)
      pvalue = as_tibble(S$coefficients)$`Pr(>|t|)`[2]
   }
    
    shapiro.test(S$residuals) -> normality
    Normality_pval = normality$p.value
    
    f_t = as_tibble(t(c(colnames(abundance_matrix)[N], pvalue, fold_change, Normality_pval)))
    colnames(f_t) = c("Taxa","P-value","Fold_change","Normality_pvalue")
    result_bugg = rbind(result_bugg, f_t)
  }
  result_bugg %>% mutate(FDR=p.adjust(`P-value`,"fdr"), FDR_normality=p.adjust(Normality_pvalue,"fdr")) -> result_bugg
  return(result_bugg)
}

```


Make absolute abundance realtive


```{r Relative abundance of pathway}
make_relative = function(Column){ return(Column/sum(Column)*100) }

```


## Data Analysis
One the functions are ready, we can start the analysis.

Let's start with some exploratory analysis.

We are interested on getting  a feel for the contamination of the data
```{r Exploration of data, echo =F }
contamination_stats()
```



And check the taxonomy distribution


```{r Exploration of data2, echo =F }
distribution_taxonomy()
```


For comparing as bars the composition, we go to the family level or above, since there are many Genuses or Species


```{r Taxon composition, echo=F}
Preprocess_f(family_matrix,"Summary_family.csv") -> family_matrix2
do_bars(family_matrix2, sample_metadata) -> List_tops
```
Which are the top families seen in each condition?
```{r Top composition}
List_tops
```


We can also look for the top Species



```{r Preprocess Species, echo = "F"}
Preprocess_f(species_matrix,"Summary_family.csv") -> species_matrix2
```

```{r Top species, include=FALSE}
do_bars(species_matrix2, sample_metadata) -> List_tops

```
```{r Top composition species}
List_tops
```


As we might be interested on looking at the top species only on pocket, we can write some code without the function.



```{r Combine Metadata and abundance matrix}
Combine_metadata = function(sample_metadata, taxa_matrix){
sample_metadata %>% filter(StudieID %in% taxa_matrix$ID) %>% arrange(StudieID) -> metadata
taxa_matrix %>% filter(ID %in% sample_metadata$StudieID) %>% arrange(ID) %>% select(-ID) -> taxa_matrix
metadata %>% mutate(N_ID=as.numeric(gsub("[^0-9.-]", "", StudieID))) %>% arrange(N_ID) -> test
metadata$StudieID = factor(metadata$StudieID,levels(factor(metadata$StudieID))[match(test$StudieID,metadata$StudieID)])
cbind(metadata,taxa_matrix) -> combined_matrix
return(combined_matrix)
}
```
```{r Look up top species in pocket}
Combine_metadata(sample_metadata=sample_metadata, taxa_matrix=species_matrix2) -> combined_matrix
taxa_matrix = select(species_matrix2,-ID)
combined_matrix %>% filter(Source=="pocket" && as.character(Group)=="0") %>% select(colnames(taxa_matrix)) %>% summarise_all(median) %>%
    gather(Taxa, Median_relative_abundance, 1:length(colnames(taxa_matrix)), factor_key=TRUE) %>%
  arrange(desc(Median_relative_abundance)) %>% head(n=5) %>% mutate(TOP="Healthy") -> top
combined_matrix %>% filter(Source=="pocket")%>% filter(Group==1) %>% select(colnames(taxa_matrix)) %>% summarise_all(median) %>%
    gather(Taxa, Median_relative_abundance, 1:length(colnames(taxa_matrix)), factor_key=TRUE) %>%
  arrange(desc(Median_relative_abundance)) %>% head(n=5) %>% mutate(TOP="Patients") -> top2
  
TOP = rbind(top,top2)
print(TOP)
```
And family prevalence on pocket
```{r bars in pocket, echo=F}
combined_matrix = Combine_metadata(sample_metadata, family_matrix2)
combined_matrix %>% filter(Source=="pocket") %>% gather(Taxa, Abundance, (dim(sample_metadata)[2]+1):dim(combined_matrix)[2], factor_key=TRUE) %>%
  ggplot() + geom_bar(aes(x=StudieID,y=Abundance, fill=Taxa), stat = "identity",position="stack") + facet_wrap(~Group, dir="v")+
   theme_bw() + theme(axis.text.x = element_text(angle = 90), legend.position = "bottom",legend.title = element_text(size = 3), legend.text = element_text(size = 6))+
    theme(legend.key.size = unit(0.1, "cm")) -> BAR1
print(BAR1)
```


### Diversity analysis
Let's check the alpha and beta diversity in the Species and Family level, and whether there are statistical differences among the groups.
```{r Diversity species}
Diversity_analysis(species_matrix2,sample_metadata)

```
```{r Diversity Family}
Diversity_analysis(family_matrix2,sample_metadata)

```


And then, look at differences on individual taxas in Species and Families.

```{r Individual differential Species abundance, echo = F}
Differential_abundance(species_matrix2,sample_metadata,"argsinsquare")

```

```{r Individual differential Family abundance, echo = F}
Differential_abundance(family_matrix2,sample_metadata,"argsinsquare")
```


Next we want to check the most abundant taxas in pocket in Patients and Controls, and the greatest median abundance fold changes

```{r Most abundant in pocket}
species_matrix2 %>% arrange(ID) %>% filter(ID %in% sample_metadata$StudieID) -> species_matrix2
sample_metadata %>% arrange(StudieID) %>% filter(StudieID %in% species_matrix2$ID) -> sample_metadata

species_matrix2 %>% mutate(Group = sample_metadata$Group, Source=sample_metadata$Source) %>% filter(Source=="pocket") -> pocket_samples
pocket_samples %>% group_by(Group) %>% summarise_if(is.numeric,median) %>%
  gather(Microorganism, Median_abundance, 2:235, factor_key=TRUE) -> Median_species
Median_species %>% ggplot(aes(x=factor(Group), y=Median_abundance)) + geom_point() +geom_violin() + theme_bw()

bigger_changes = tibble()
for (i in unique(Median_species$Microorganism)){
  Median_species %>% filter(Microorganism==i) -> M
  filter(M, Group==1)$Median_abundance/filter(M, Group==0)$Median_abundance -> Change
  ti = tibble(Micro=M$Microorganism[1], Log_fold_Change=log(Change))
  bigger_changes = rbind(bigger_changes,ti)
}
bigger_changes %>% arrange(desc(Log_fold_Change)) -> Fold_changes


Median_species %>% filter(Group==0) %>% arrange(desc(Median_abundance)) -> species_control
Median_species %>% filter(Group==1) %>% arrange(desc(Median_abundance)) -> species_patient
```

Most abundant species in control samples
```{r Most abundant in pocket2}
print(species_control)
```
And in Patients
```{r Most abundant in pocket3}
species_patient
```
Biggest fold changes
```{r Most abundant in pocket4}
Fold_changes
```



We are also interested in the differences between patients and controls in two specific organisims, namely Porphyromonas_gingivalis and Aggregatibacter_actinomycetemcomitans

```{r p.g}
pocket_samples %>% select(s__Porphyromonas_gingivalis, ID, Group) -> por_ging_pocket
ggplot(por_ging_pocket,aes(x=factor(Group),y=s__Porphyromonas_gingivalis)) + geom_boxplot() + geom_point() + theme_bw()
wilcox.test(filter(por_ging_pocket, Group==1)$s__Porphyromonas_gingivalis,filter(por_ging_pocket, Group==0)$s__Porphyromonas_gingivalis)
```
```{r a.a}
pocket_samples %>% select(s__Aggregatibacter_actinomycetemcomitans, ID, Group) -> aa_pocket
ggplot(aa_pocket,aes(x=factor(Group),y=s__Aggregatibacter_actinomycetemcomitans)) + geom_boxplot() + geom_point() + theme_bw()

wilcox.test(filter(aa_pocket, Group==1)$s__Aggregatibacter_actinomycetemcomitans,filter(aa_pocket, Group==0)$s__Aggregatibacter_actinomycetemcomitans)

```


Finally, we will check Pathway abundance changes between the two conditions.

First we preprocess the data
```{r Pathway preprocessing, echo=F,include=FALSE, warnings=F}
new_i = vector()
for (i in colnames(pathway_abundance)){
  if (i=="# Pathway"){ 
    i = "Pathway"
  }else{
    i = str_replace(string = i,pattern = "_merged_Abundance",replacement = "")
  }
  new_i = c(new_i,i)
}

colnames(pathway_abundance) = new_i
pathway_abundance %>% select(-Pipelines) -> pathway_abundance
pathway_abundance %>% mutate_if(is_numeric,make_relative) -> pathway_abundance

new_i = colnames(pathway_abundance)
Pathway = pathway_abundance$Pathway
pathway_abundance %>% select(-Pathway) %>% t() -> pathway_abundance2

colnames(pathway_abundance2) = Pathway
pathway_abundance2 %>% as_tibble() %>% mutate(ID = new_i[2:length(new_i)]) -> pathway_abundance2
pathway_abundance2 %>% select(-c(UNMAPPED,UNINTEGRATED)) -> pathway_abundance2
pathway_abundance2 %>% select(c("ID",colnames(select(pathway_abundance2,-ID)))) -> pathway_abundance2

```

Then we test for changes on pathway abundance

```{r Differential pathway abundance results, echo=F, warning=F}
differential_pathway_analysis(pathway_abundance2,sample_metadata) -> pathway_results
```
```{r Pathway abundance output}
print(pathway_results)
```
