## Metabolite - Abundance tests
In order to use TMAOmetabolites_16.r, four files should be provided.  
1. 16S count table. Please follow https://github.com/alexa-kur/miQTL_cookbook#Appendix2 until generating the taxonomy_table.txt in step 1.2  
2. Linking_table. This is a two column file, where the first column are Samples ID and the second are sequencing IDs that are be used on teh 16S count table.  
3. Covariates file. Should include an ID column (Samples ID) and a column for Age, Gender and BMI.  
4. Phenotype file. Should include an ID column (Samples ID) and a column for each dependent variable to be fitted into a model.  

Calling the script:  

Rscript TMAOmetabolites_16.r $1 $2 $3 $4  

Where each of the $number representes each of the files described above.  


### Process  
All samples from the four files are matched based on the ID column (SampleID in the 16S count table).  
If Linking_Table has more than one sequencing ID per Sample ID (technical replicates), a random one will be chosen.  
All unclassified reads are filtered from the 16S count table.  
Then, there will be an iteration through the taxonomies in each taxonomic label (Genus, Family, Order, Class, Phylum) in the 16S count table. Then, each taxonomic level will be divided in Males and Females, along with a compelte set, according to the Gender values from the Covariate file.
The counts in a taxonomic level, for each Gender group, will be transformed using clr from the Microbiome R package (a pseudocount of half the value of the lowest count in each taxon is added before transformation).  
In each of the transformed count data, a linear model will be fitted in each of the columns of the Phenotype file, except the first where ID should be included.  
Each of the models will look like:  

Inverse(rank(Metabolite)) ~ Abundance_Bug + Gender + BMI + Age  

FDR values are generated by producing a distribution of Pvalues from 100 permutations of the abundance data. Then, for a given Pvalue generated from the real data, the amount of Pvalues lower or equal than that in the random distribution of Pvalues is counted (FP) and the number of Pvalues lower and equal tahnt that one in the real Pvalue distribution (P).  

FDR is then calculated according to:  

FDR = (FP/100)/P  

A file is produced per each gender and taxonomic level. Each of them contain results for all the Taxons in each Phenotype.

